[ONNX] Set shape/type into torchscript

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96350
* #96879
* __->__ #96349


Fixes https://github.com/pytorch/pytorch/pull/95676#issuecomment-1460588229

PS: It doesn't seem the exported ONNX_proto having type now. I wonder if there was a ONNX pass doing this for us (converting torch dtype to onnx dtype during exporting.)

Type promotion issue would be raised with an error if we want to set type
```python
onnxscript_value.dtype = expected_value.dtype
```
onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] Shape inference error(s): (op_type:aten_add, node name: aten_add_1): [ShapeInferenceError] (op_type:Add, node name: n3): B has inconsistent type tensor(int64)

titaiwangms committed:
[ONNX] Set shape/type into torchscript

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Set shape/type into torchscript"



Fixes https://github.com/pytorch/pytorch/pull/95676#issuecomment-1460588229

PS: It doesn't seem the exported ONNX_proto having type now. I wonder if there was a ONNX pass doing this for us (converting torch dtype to onnx dtype during exporting.)

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/fx/exporter.py:
```diff
@@ -181,13 +181,25 @@ def _fill_tensor_meta(
         # Only set shape for now as we don't need type information.
+        # import pdb; pdb.set_trace()
         onnxscript_value.shape = tuple(expected_value.size())
+        onnxscript_value.type = expected_value.dtype
```
I would try just setting shape for now because we want to be able to catch type mismatches I think? onnx infer shape will fill in the type info for us. So far it works fine

titaiwangms replied:
I am not sure if @wschin need this for custom Op

justinchuby replied:
ok that's right. Possible to specify if we want to add dtype then? So we only assign dtypes to custom ops

justinchuby replied:
In the future I am hoping all ops will have their opschema defined.

titaiwangms replied:
It seems setting dtype would break graph somehow...

titaiwangms committed:
Update on "[ONNX] Set shape/type into torchscript"



Fixes https://github.com/pytorch/pytorch/pull/95676#issuecomment-1460588229

PS: It doesn't seem the exported ONNX_proto having type now. I wonder if there was a ONNX pass doing this for us (converting torch dtype to onnx dtype during exporting.)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Set shape/type into torchscript"



Fixes https://github.com/pytorch/pytorch/pull/95676#issuecomment-1460588229

PS: It doesn't seem the exported ONNX_proto having type now. I wonder if there was a ONNX pass doing this for us (converting torch dtype to onnx dtype during exporting.)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Set shape/type into torchscript"



Fixes https://github.com/pytorch/pytorch/pull/95676#issuecomment-1460588229

PS: It doesn't seem the exported ONNX_proto having type now. I wonder if there was a ONNX pass doing this for us (converting torch dtype to onnx dtype during exporting.)

Type promotion issue would be raised with an error if we want to set type
```python
onnxscript_value.dtype = expected_value.dtype
```
onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] Shape inference error(s): (op_type:aten_add, node name: aten_add_1): [ShapeInferenceError] (op_type:Add, node name: n3): B has inconsistent type tensor(int64)

[ghstack-poisoned]

titaiwangms commented:
cc @BowenBao @justinchuby @wschin

titaiwangms commented:
@pytorchbot merge -g

BowenBao commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -163,13 +163,26 @@ def _fill_tensor_meta(
         # Only set shape for now as we don't need type information.
         onnxscript_value.shape = tuple(expected_value.size())
+        # FIXME(titaiwang): type promotion is not supported yet,
+        # so set dtype will break onnx shape inference.
```
Are we postponing the error?

titaiwangms replied:
when type_promotion is provided, this will be fixed. Currently, it's blocked by aten_add. We could add `CastLike(self, other)` to pass CI test which would be a workaround.

titaiwangms replied:
dtype enabled or not is not blocking anything though. It's nice to have.

titaiwangms replied:
I intend to enable it in the next PR with dynamic support.

titaiwangms replied:
@BowenBao It's enabled as type_promotion error is ignored https://github.com/microsoft/onnx-script/blob/ccf55862b55ccc4ecff4b8d4cc447c8db931804f/onnxscript/function_libs/torch_aten/graph_building.py#L488

So far, ORT is happy with it.

titaiwangms commented:
@pytorchmergebot merge -g

titaiwangms commented:
@pytorchbot rebase

titaiwangms committed:
Update on "[ONNX] Set shape/type into torchscript"



Fixes https://github.com/pytorch/pytorch/pull/95676#issuecomment-1460588229

PS: It doesn't seem the exported ONNX_proto having type now. I wonder if there was a ONNX pass doing this for us (converting torch dtype to onnx dtype during exporting.)

Type promotion issue would be raised with an error if we want to set type
```python
onnxscript_value.dtype = expected_value.dtype
```
onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] Shape inference error(s): (op_type:aten_add, node name: aten_add_1): [ShapeInferenceError] (op_type:Add, node name: n3): B has inconsistent type tensor(int64)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Set shape/type into torchscript"



Fixes https://github.com/pytorch/pytorch/pull/95676#issuecomment-1460588229

PS: It doesn't seem the exported ONNX_proto having type now. I wonder if there was a ONNX pass doing this for us (converting torch dtype to onnx dtype during exporting.)

Type promotion issue would be raised with an error if we want to set type
```python
onnxscript_value.dtype = expected_value.dtype
```
onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] Shape inference error(s): (op_type:aten_add, node name: aten_add_1): [ShapeInferenceError] (op_type:Add, node name: n3): B has inconsistent type tensor(int64)

[ghstack-poisoned]

titaiwangms commented:
Need https://github.com/pytorch/pytorch/pull/96793

titaiwangms committed:
Update on "[ONNX] Set shape/type into torchscript"



Fixes https://github.com/pytorch/pytorch/pull/95676#issuecomment-1460588229

PS: It doesn't seem the exported ONNX_proto having type now. I wonder if there was a ONNX pass doing this for us (converting torch dtype to onnx dtype during exporting.)

Type promotion issue would be raised with an error if we want to set type
```python
onnxscript_value.dtype = expected_value.dtype
```
onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] Shape inference error(s): (op_type:aten_add, node name: aten_add_1): [ShapeInferenceError] (op_type:Add, node name: n3): B has inconsistent type tensor(int64)

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Set shape/type into torchscript"



Fixes https://github.com/pytorch/pytorch/pull/95676#issuecomment-1460588229

PS: It doesn't seem the exported ONNX_proto having type now. I wonder if there was a ONNX pass doing this for us (converting torch dtype to onnx dtype during exporting.)

Type promotion issue would be raised with an error if we want to set type
```python
onnxscript_value.dtype = expected_value.dtype
```
onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] Shape inference error(s): (op_type:aten_add, node name: aten_add_1): [ShapeInferenceError] (op_type:Add, node name: n3): B has inconsistent type tensor(int64)

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -163,13 +163,26 @@ def _fill_tensor_meta(
+        # TODO(titaiwang): Support List of Tensors: https://github.com/microsoft/onnx-script/issues/481
+        # This is the case that fx has [tensor, tensor, ...], but onnxscript_value wrapped it as a single tensor.
+        # graph_buiding not yet support list of tensors, so we don't need to handle this case for now.
+        # eg: aten_split
```
@justinchuby related https://github.com/microsoft/onnx-script/pull/528

titaiwangms commented:
@pytorchbot merge -f "unrelated inductor tests failed on macos-12-py3-arm64"
